src/signing/collector/signatures_collector_preprocessor.rs

```diff
@@ -0,0 +1,118 @@
+use crate::prelude::*;
+
+pub struct SignaturesCollectorPreprocessor {
+    transactions: IndexSet<TXToSign>,
+}
+
+pub fn sort_group_factors(
+    used_factor_sources: HashSet<HDFactorSource>,
+) -> IndexSet<FactorSourcesOfKind> {
+    let factors_of_kind = used_factor_sources
+        .into_iter()
+        .into_grouping_map_by(|x| x.factor_source_kind())
+        .collect::<IndexSet<HDFactorSource>>();
+
+    let mut factors_of_kind = factors_of_kind
+        .into_iter()
+        .map(|(k, v)| (k, v.into_iter().sorted().collect::<IndexSet<_>>()))
+        .collect::<IndexMap<FactorSourceKind, IndexSet<HDFactorSource>>>();
+
+    factors_of_kind.sort_keys();
+
+    factors_of_kind
+        .into_iter()
+        .map(|(k, v)| FactorSourcesOfKind::new(k, v).unwrap())
+        .collect::<IndexSet<_>>()
+}
+
+impl SignaturesCollectorPreprocessor {
+    pub(super) fn new(transactions: IndexSet<TXToSign>) -> Self {
+        Self { transactions }
+    }
+
+    pub(super) fn preprocess(
+        self,
+        all_factor_sources_in_profile: IndexSet<HDFactorSource>,
+    ) -> (Petitions, IndexSet<FactorSourcesOfKind>) {
+        let transactions = self.transactions;
+        let mut petitions_for_all_transactions = IndexMap::<IntentHash, PetitionTransaction>::new();
+
+        let all_factor_sources_in_profile = all_factor_sources_in_profile
+            .into_iter()
+            .map(|f| (f.factor_source_id(), f))
+            .collect::<HashMap<FactorSourceIDFromHash, HDFactorSource>>();
+
+        let mut factor_to_payloads = HashMap::<FactorSourceIDFromHash, IndexSet<IntentHash>>::new();
+
+        let mut used_factor_sources = HashSet::<HDFactorSource>::new();
+
+        let mut use_factor_in_tx = |id: &FactorSourceIDFromHash, txid: &IntentHash| {
+            if let Some(ref mut txids) = factor_to_payloads.get_mut(id) {
+                txids.insert(txid.clone());
+            } else {
+                factor_to_payloads.insert(*id, IndexSet::from_iter([txid.clone()]));
+            }
+
+            assert!(!factor_to_payloads.is_empty());
+
+            let factor_source = all_factor_sources_in_profile
+                .get(id)
+                .expect("Should have all factor sources");
+            used_factor_sources.insert(factor_source.clone());
+
+            assert!(!used_factor_sources.is_empty());
```

Comment:
Not sure what we gain with this - validate that rust HashSet works as expected?

Reply:
(remnant of old code)
TODO Remove